src/lib/y2storage/storage_class_wrapper.rb

```diff
@@ -0,0 +1,258 @@
+      # instance ones.
+      #
+      # @see #storage_forward
+      def storage_class_forward(method, as: nil, raise_errors: [])
+        modifiers = { as: as, raise_errors: raise_errors }
+        define_singleton_method(method) do |*args|
+          StorageClassWrapper.forward(storage_class, method, modifiers, *args)
+        end
+      end
+
+      def storage_class
+        @storage_class
+      end
+
+      def storage_class_name
+        @storage_class_name ||= storage_class.name.split("::").last
+      end
+
+      def storage_class_underscored_name
+        @storage_class_underscored_name ||= StorageClassWrapper.underscore(storage_class_name)
+      end
+
+      # Alternative constructor used internally by this module in order to get
+      # fully downcasted objects.
+      #
+      # @param storage object to be wrapped
+      def downcasted_new(object)
+        @downcast_class_names.each do |class_name|
+          underscored = StorageClassWrapper.underscore(class_name.split("::").last)
+          check_method = :"#{underscored}?"
+          cast_method = :"to_#{underscored}"
+          next unless Storage.send(check_method, object)
+
+          klass = StorageClassWrapper.class_for(class_name)
+          return klass.downcasted_new(Storage.send(cast_method, object))
+        end
+        new(object)
+      end
+    end
+
+    # @see ClassMethods#storage_forward
+    def self.forward(storage_object, method, modifiers, *args)
+      wrapper_class_name = modifiers[:as]
+      raise_errors = modifiers[:raise_errors] || []
+
+      processed_args = processed_storage_args(*args)
+      result = storage_object.send(method, *processed_args)
+      processed_storage_result(result, class_for(wrapper_class_name))
+    rescue Storage::WrongNumberOfChildren
+      raise_errors.include?("WrongNumberOfChildren") ? raise : nil
+    rescue Storage::DeviceHasWrongType
+      raise_errors.include?("DeviceHasWrongType") ? raise : nil
+    end
+
+    def self.class_for(class_name)
+      class_name ? Y2Storage.const_get(class_name) : nil
+    end
+
+    def self.processed_storage_args(*args)
+      args.map { |arg| arg.respond_to?(:to_storage_value) ? arg.to_storage_value : arg }
+    end
+
+    def self.processed_storage_result(result, wrapper_class)
+      result = result.to_a if result.class.name.start_with?("Storage::Vector")
+
+      return result unless wrapper_class
+
+      if result.is_a?(Array)
+        result.map { |o| object_for(wrapper_class, o) }
+      else
+        object_for(wrapper_class, result)
+      end
+    end
+
+    def self.underscore(camel_case_name)
+      camel_case_name.gsub(/(.)([A-Z])/, '\1_\2').downcase
+    end
+
+    def self.object_for(wrapper_class, storage_object)
+      if wrapper_class.respond_to?(:downcasted_new)
+        wrapper_class.downcasted_new(storage_object)
+      else
+        wrapper_class.new(storage_object)
+      end
+    end
+  end
```

Comment:
In general I do not like number of methods in this module. Problem is that it add bunch of public available methods, which looks wrong for me. I think we need some encapsulation here.